src/hooks/useUniversalRouter.ts

```diff
@@ -0,0 +1,87 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { TransactionResponse } from '@ethersproject/providers'
+import { Percent } from '@uniswap/sdk-core'
+import { SwapRouter, UNIVERSAL_ROUTER_ADDRESS } from '@uniswap/universal-router-sdk'
+import { FeeOptions } from '@uniswap/v3-sdk'
+import { useWeb3React } from '@web3-react/core'
+import { useCallback } from 'react'
+import { InterfaceTrade } from 'state/routing/types'
+import { calculateGasMargin } from 'utils/calculateGasMargin'
+import isZero from 'utils/isZero'
+import { swapErrorToUserReadableMessage } from 'utils/swapErrorToUserReadableMessage'
+
+import { PermitSignature } from './usePermitAllowance'
+
+interface SwapOptions {
+  slippageTolerance: Percent
+  deadline?: BigNumber
+  permit?: PermitSignature
+  feeOptions?: FeeOptions
+}
+
+export function useUniversalRouterSwapCallback(trade: InterfaceTrade | undefined, options: SwapOptions) {
+  const { account, chainId, provider } = useWeb3React()
+
+  return useCallback(async (): Promise<TransactionResponse> => {
+    try {
+      if (!account || !chainId || !provider || !trade) throw new Error('missing input')
+
+      const calldata = SwapRouter.swapERC20CallParameters(trade, {
+        slippageTolerance: options.slippageTolerance,
+        deadlineOrPreviousBlockhash: options.deadline?.toString(),
+        inputTokenPermit: options.permit,
+        fee: options.feeOptions,
+      })
+      const tx =
+        calldata.value && !isZero(calldata.value)
+          ? {
+              from: account,
+              to: UNIVERSAL_ROUTER_ADDRESS(chainId),
```

Comment:
this will throw on unsupported chains--are you expecting that/handling it the way you'd like?

Reply:
I've added a check to useSyncFlags.ts#usePermit2 so that permit2 will only be enabled on chains on which the Universal Router is supported.